Source/pack.cpp

```diff
@@ -68,34 +87,108 @@ void PackNetItem(const Item &item, ItemNetPack &packedItem)
+
+	if (isPregenItem && hasMultipleFlags(iCreateInfo))
+		return false;
+	if (isGroundItem && (iCreateInfo & ~CF_USEFUL) != 0)
+		return false;
+	if (isTownItem && hasMultipleFlags(iCreateInfo))
+		return false;
+	return true;
 }
 
-template <typename T>
-void LogFailedJoinAttempt(const char *condition, const char *name, T value)
+bool IsTownItemValid(uint16_t iCreateInfo)
 {
-	LogDebug("Remote player validation failed: ValidateField({}: {}, {})", name, value, condition);
+	const uint8_t level = iCreateInfo & CF_LEVEL;
+	const bool isBoyItem = (iCreateInfo & CF_BOY) != 0;
+
+	if (isBoyItem && level <= MaxCharacterLevel)
+		return true;
+
+	return level <= 30;
 }
 
-template <typename T1, typename T2>
-void LogFailedJoinAttempt(const char *condition, const char *name1, T1 value1, const char *name2, T2 value2)
+bool IsUniqueMonsterItemValid(uint16_t iCreateInfo, uint32_t dwBuff)
 {
-	LogDebug("Remote player validation failed: ValidateFields({}: {}, {}: {}, {})", name1, value1, name2, value2, condition);
+	const uint8_t level = iCreateInfo & CF_LEVEL;
+	const bool isHellfireItem = (dwBuff & CF_HELLFIRE) != 0;
+
+	for (int i = 0; UniqueMonstersData[i].mName != nullptr; i++) {
+		const auto &uniqueMonsterData = UniqueMonstersData[i];
+		const auto &uniqueMonsterLevel = static_cast<uint8_t>(MonstersData[uniqueMonsterData.mtype].level);
+
+		if (!isHellfireItem && IsAnyOf(uniqueMonsterData.mtype, MT_HORKDMN, MT_DEFILER, MT_NAKRUL)) {
+			// These monsters don't appear in Diablo
+			continue;
+		}
+
+		if (level == uniqueMonsterLevel) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+bool IsDungeonItemValid(uint16_t iCreateInfo, uint32_t dwBuff)
+{
+	const uint8_t level = iCreateInfo & CF_LEVEL;
+	const bool isHellfireItem = (dwBuff & CF_HELLFIRE) != 0;
+
+	for (int i = 0; MonstersData[i].name != nullptr; i++) {
+		const auto &monsterData = MonstersData[i];
+		auto monsterLevel = static_cast<uint8_t>(monsterData.level);
+
+		if (IsAnyOf(i, MT_HORKDMN, MT_DEFILER, MT_NAKRUL)) {
+			continue;
+		}
+
+		if (i == MT_DIABLO && !isHellfireItem) {
+			monsterLevel -= 15;
+		}
+
+		if (level == monsterLevel) {
+			return true;
+		}
+	}
+
+	return level <= 30;
+}
+
+bool UnPackNetItem(const Player &player, const ItemNetPack &packedItem, Item &item)
+{
+	item = {};
+	_item_indexes idx = static_cast<_item_indexes>(SDL_SwapLE16(packedItem.def.wIndx));
+	if (idx < 0 || idx > IDI_LAST)
+		return false;
+	if (idx == IDI_EAR) {
+		RecreateEar(item, SDL_SwapLE16(packedItem.ear.wCI), SDL_SwapLE32(packedItem.ear.dwSeed), packedItem.ear.bCursval, packedItem.ear.heroname);
+		return true;
+	}
+
+	uint16_t creationFlags = SDL_SwapLE16(packedItem.item.wCI);
+	uint32_t dwBuff = SDL_SwapLE16(packedItem.item.dwBuff);
+	ValidateField(creationFlags, IsCreationFlagComboValid(creationFlags));
+	if ((creationFlags & CF_TOWN) != 0)
+		ValidateField(creationFlags, IsTownItemValid(creationFlags));
+	else if ((creationFlags & CF_UPER15) != 0 && (creationFlags & CF_USEFUL) != CF_USEFUL)

```

Comment:
Maybe this is more readable?
```suggestion
	else if ((creationFlags & (CF_UPER1 | CF_UPER15)) == CF_UPER15)
```

Reply:
I like having the two separate checks because it seems to more naturally read, "CF_UPER15 is set and CF_USEFUL is not". However, it's difficult to understand what's happening here without realizing that `CF_USEFUL` is a mask that includes `CF_UPER15` regardless so I'd be okay with the second suggestion as well.

Reply:
For completeness here is my first comment in code form:
```suggestion
	else if ((creationFlags & CF_UPER15) != 0 && (creationFlags & CF_UPER1) == 0)
```